UPDATE_APPEARANCE PROFILING TESTMERGE

LemonInTheDark opened the pull request:
## About The Pull Request

[Saves 0.2 seconds of init time. 50% of emissive blockers](https://github.com/tgstation/tgstation/commit/8318b648f6d32844aacbfb4c309152cd45801f5c)

Emissive blockers are a decent expense during init, even these, which are the ones that update outside of initialize.
I've inlined them, removed some redundant vars and checks, reduced the arg count, and shifted some things around. This ends up saving 200ms, or 50% of its total cost.

I also shifted mutable_appearance about a bit. it's not a massive saving, but it is technically faster

[Prevents a few redundant appearance_updates, saves 0.8 seconds of init](https://github.com/tgstation/tgstation/commit/5475cd778b66b22b1e2c8d86b2c6d59fb84f219a)

Prequisit info: update_appearance is decently expensive
It's good then to only do it if we have a reason to, right?

Me and moth were shooting the shit about just general init time, and we came up with the idea of tracking which update_appearances actually "worked" and which didn't.

That bit comes later, let's enjoy the fruits of that work first

First, holograms were calling update_appearance on process, for almost no reason. 
I patched the one event they don't already "react" to, and then locked it behind a change decection if. 
good for live, doesn't impact init.

Next, decals. If you add a decal to something before it inits, it'll react to the after successful init signal. 
The trouble is the same atom could have its appearance updated from this MORE then once, since decals can be stacked on tiles, and signal unregisters don't work once the signal is sent.
So we add a flag to track if we've had this happen to us or not, so it only happens once. 
saves 80 ms

Power! lots of things call power_change on init, often more then once. 
We'll update appearance for each of those calls, even if only one is an actual change.
That's silly, better to track what sort of power we're using for our appearance and go off that changing

This was taking about 300ms. Really stupid

Icon smoothing. After emissive blockers were added, any change to something's icon smoothing would lead to an update_appearance call.
Nasty shit, specially cause of walls, which don't even use emissive blockers.
Ok then, so we'll always update appearance for movables, and will allow turfs that are interested to hook it manually. 
Not many of those anyhow
This is slightly a dev ux thing, but it saves 600ms so I think it's worth it. Rare case anyway

Telecomms:
telecomm machines were updating appearance on process. This is to cover for them turning on/off on process.
Better then to just check if on actually changed. 
This cost adds up midgame, doesn't impact init tho

Materials:
There's this update_appearance call in material on_apply. it doesn't do anything.
The logs will lie to you and say it does, but it's just like reapplying emissives. It doesn't need to exist
Saves like 50ms

[Uses defines to wrap update_appearance for tracking](https://github.com/tgstation/tgstation/commit/4fa82e1c9d93577aadb3c743f17196331f62e67c)

I need file and line number for my tracking, so I need to override update_appearance calls, and also preferably not require every override of update_appearance to handle dummy file + line args.

So instead, I created a wrapper proc that checks to see if appearanaces match (they're unique remember, the two of the same visual appearance will be equivalent)
The trouble is I can't intercept JUST proc calls, or JUST function definitions with defines. it needs to be both.

So I renamed the /update_appearance proc to /_update_appearance

this way I can capture old uses, and don't need to worry about merge/dev brain skew

It does mean that all update_appearance proc definitions now look weird tho. 
My profiling is leaking into dev ux. I wish I had better templating.

## Why It's Good For The Game

Saves init time, saves machine cpu usage, allows for appearance profiling in future

LemonInTheDark committed:
Saves 0.2 seconds of init time. 50% of emissive blockers

Emissive blockers are a decent expense during init, even these, which
are the ones that update outside of initialize.
I've inlined them, removed some redundant vars and checks, reduced the arg count,
and shifted some things around. This ends up saving 200ms, or 50% of its
total cost.

I also shifted mutable_appearance about a bit. it's not a massive
saving, but it is technically faster

LemonInTheDark committed:
Prevents a few redundant appearance_updates, saves 0.8 seconds of init

Prequisit info: update_appearance is decently expensive
It's good then to only do it if we have a reason to, right?

Me and moth were shooting the shit about just general init time, and we
came up with the idea of tracking which update_appearances actually
"worked" and which didn't.

That bit comes later, let's enjoy the fruits of that work first

First, holograms were calling update_appearance on process, for almost
no reason. I patched the one event they don't already "react" to, and
then locked it behind a change decection if. good for live, doesn't
impact init.

Next, decals. If you add a decal to something before it inits, it'll
react to the after successful init signal. The trouble is the same atom
could have its appearance updated from this MORE then once, since decals
can be stacked on tiles, and signal unregisters don't work once the
signal is sent.
So we add a flag to track if we've had this happen to us or not, so it
only happens once. saves 80 ms

Power! lots of things call power_change on init, often more then once.
We'll update appearance for each of those calls, even if only one is an actual change.
That's silly, better to track what sort of power we're using for our
appearance and go off that changing

This was taking about 300ms. Really stupid

Icon smoothing. After emissive blockers were added, any change to
something's icon smoothing would lead to an update_appearance call.
Nasty shit, specially cause of walls, which don't even use emissive
blockers.

Ok then, so we'll always update appearance for movables, and will
allow turfs that are interested to hook it manually. Not many of those
anyhow
This is slightly a dev ux thing, but it saves 600ms so I think it's
worth it. Rare case anyway

Telecomms:
telecomm machines were updating appearance on process. This is to cover
for them turning on/off on process.
Better then to just check if on actually changed. This cost adds up
midgame, doesn't impact init tho

Materials:
There's this update_appearance call in material on_apply. it doesn't do
anything.
The logs will lie to you and say it does, but it's just like reapplying
emissives. It doesn't need to exist

LemonInTheDark committed:
Uses defines to wrap update_appearance for tracking

I need file and line number for my tracking, so I need to override
update_appearance calls, and also preferably not require every override
of update_appearance to handle dummy file + line args.

So instead, I created a wrapper proc that checks to see if appearanaces
match (they're unique remember, the two of the same visual appearance will be equivilant)
The trouble is I can't intercpet JUST proc calls, or JUST function
definitions with defines. it needs to be both.

So I renamed the /update_appearance proc to /_update_appearance

this way I can capture old uses, and don't need to worry about merge/dev
brain squew

It does mean that all update_appearance proc definitions now look weird
tho. My profiling is leaking into dev ux. I wish I had better
templating.

Mothblocks commented:
We probably should not be merging this if you have to `_update_appearance`, can we just test merge this and get intel instead?

Mothblocks commented on code/__HELPERS/stat_tracking.dm:
```diff
@@ -25,3 +27,15 @@
+
+	output += "key, cost, count"
+	for (var/key in costs)
+		output += "[key], [costs[key]], [counts[key]]"
```
You should probably validate (just throwing an exception is fine) that `key` doesn't have any commas

LemonInTheDark replied:
gonna just strip all commas, see new pr

Mothblocks commented:
Idea: Include the appearance tracking code in master today, but leave it completely unused.

Include a comment explaining the grep you need to find and replace with in order to make it work, and what defines to flip, etc.

That way we can do this again in a year without needing to review the code as finely.

LemonInTheDark commented:
> Idea: Include the appearance tracking code in master today, but leave it completely unused.
> 
> Include a comment explaining the grep you need to find and replace with in order to make it work, and what defines to flip, etc.
> 
> That way we can do this again in a year without needing to review the code as finely.

I'm down, sounds like it'd work ok without interrupting dev ux

Mothblocks commented:
Let's make a PR just for that, and maybe combine it with the emissive blocker micro-opts, so we have something to merge into master. Then you can keep this PR up with a DNM.

LemonInTheDark commented:
> Screenshot tests failed!
> 
> [Download zip file of new screenshots.](https://file.house/WXN8.zip)
> ## Diffs
> See snapshot diffs
> Name 	Expected image 	Produced image 	Diff
> screenshot_antag_icons_abductor 	![](https://camo.githubusercontent.com/48ca9a285574eb4d3f5f17e30f8a10903e16805462f7d6f71d0295f716c3e781/68747470733a2f2f66696c652e686f7573652f623249742e706e67) 	![](https://camo.githubusercontent.com/c5cdeb21a1f2e9e9628a59c59a8d1c727335ecb420ddf901ffd0c3c34c9ee3bc/68747470733a2f2f66696c652e686f7573652f425a6c662e706e67) 	![](https://camo.githubusercontent.com/bc575c380b5fdc9ed63d282ecfcf96860d654532f1edc31ee6688bb36eb31336/68747470733a2f2f66696c652e686f7573652f446b42682e706e67)
> screenshot_antag_icons_blobinfection 	![](https://camo.githubusercontent.com/5c2822663dc696cd6cc9aa8295be2aaddd95c402291501dcd6728790676233ed/68747470733a2f2f66696c652e686f7573652f327376582e706e67) 	![](https://camo.githubusercontent.com/7edc805d08c2f44d2303a77c95dc7d11cb942d3e1a1cda6331e7b094272011b3/68747470733a2f2f66696c652e686f7573652f756769552e706e67) 	![](https://camo.githubusercontent.com/5a18ce5342a1fb3883990f101973c245f4cb85c998fc860377778885205eb586/68747470733a2f2f66696c652e686f7573652f72795f522e706e67)
> screenshot_antag_icons_bloodbrother 	![](https://camo.githubusercontent.com/f1486e72154017f13408996ea5bbce0c4c83537e729dbdc025c426afbd244f60/68747470733a2f2f66696c652e686f7573652f734d554b2e706e67) 	![](https://camo.githubusercontent.com/268677571d2523a9c180d2fa2db7938042f6ea88dc7b6b556c017ba8fe55d5b8/68747470733a2f2f66696c652e686f7573652f416939342e706e67) 	![](https://camo.githubusercontent.com/f744735094f00830da7634235a44a62d5b261832648f5e2602787e0065c6ea89/68747470733a2f2f66696c652e686f7573652f2d4b69632e706e67)
> screenshot_antag_icons_changeling 	![](https://camo.githubusercontent.com/d4ace84291f6c028ad826bcc59fb31a53dbb1aba79580cee9b2a035533d92ede/68747470733a2f2f66696c652e686f7573652f526f6b632e706e67) 	![](https://camo.githubusercontent.com/d74e541acc517d39883aa9aaa1fbc409d9e8e0d2a210261e8ca20eda4e8d4850/68747470733a2f2f66696c652e686f7573652f4b6976502e706e67) 	![](https://camo.githubusercontent.com/bd1963ae13cf678cfb3ef76e34cdea65bc41770a5972cd170730a0ec98051be7/68747470733a2f2f66696c652e686f7573652f444862522e706e67)
> screenshot_antag_icons_clownoperative 	![](https://camo.githubusercontent.com/c9e670685d381246c454a6c403ed9bfc63c5bc7be6004146662f3d450a96bce5/68747470733a2f2f66696c652e686f7573652f2d4870352e706e67) 	![](https://camo.githubusercontent.com/d037e3b86911970f3a69300094f6f419a7da37a38135b5b51c748ad24011b157/68747470733a2f2f66696c652e686f7573652f576d59432e706e67) 	![](https://camo.githubusercontent.com/1f90d208eab2ecef91498542db8489174be9a3991d4f127cdddb155931ea494f/68747470733a2f2f66696c652e686f7573652f585248492e706e67)
> screenshot_antag_icons_cultist 	![](https://camo.githubusercontent.com/c236f235ae7af182b8a7b13bca379203b1f525df7cd0092ef65bc8e52871e1d5/68747470733a2f2f66696c652e686f7573652f756b30302e706e67) 	![](https://camo.githubusercontent.com/e19cb829cd442a14b85ee5b4da7a158e3d8e2a82b6e65b5d792efeccf55400e0/68747470733a2f2f66696c652e686f7573652f565870482e706e67) 	![](https://camo.githubusercontent.com/7a5d602ebe8affb5266e82f31d72bfe47cefc59f6dc6e549cf47f07d91c91a96/68747470733a2f2f66696c652e686f7573652f59446f372e706e67)
> screenshot_antag_icons_fugitive 	![](https://camo.githubusercontent.com/25611cd9fb9fc5346974865e2e3e143e77ddf73aba3184f14dabfde776bb6267/68747470733a2f2f66696c652e686f7573652f674444672e706e67) 	![](https://camo.githubusercontent.com/f3fb4e1864e82817dd13f748cc1f2ed9c98c379725fad974cbf9803ce65fe99d/68747470733a2f2f66696c652e686f7573652f517a73782e706e67) 	![](https://camo.githubusercontent.com/224fd933a62654ab0b89b5d84b3771e501b95914d43410bf770e42a8fd28e627/68747470733a2f2f66696c652e686f7573652f53536d572e706e67)
> screenshot_antag_icons_headrevolutionary 	![](https://camo.githubusercontent.com/fd99a78aba94ede6faeb762db4b4b0542a28fa841123af6c0cb3ba38428edf70/68747470733a2f2f66696c652e686f7573652f69376b482e706e67) 	![](https://camo.githubusercontent.com/84690e6e6d568fd545e97a5831592e219d2ebb2bedfe82b125e1405136ef6d29/68747470733a2f2f66696c652e686f7573652f4a755a772e706e67) 	![](https://camo.githubusercontent.com/9a9a7392191d9043da05160cd800b84bf0432db95fd52afe6232900b2b7d6ec4/68747470733a2f2f66696c652e686f7573652f315568712e706e67)
> screenshot_antag_icons_heretic 	![](https://camo.githubusercontent.com/64f22a8166e3b1a3cf4f1c6fe9ff0bfd3d5dfe9ce51e8df348cfe28c7c3ff0d8/68747470733a2f2f66696c652e686f7573652f6a4d6f332e706e67) 	![](https://camo.githubusercontent.com/fa9fb8ee5c93e481b1c04c8db7d1c210c0a4342faa7a710939124d86c7f678eb/68747470733a2f2f66696c652e686f7573652f537673542e706e67) 	![](https://camo.githubusercontent.com/d5f74f459d77953e795f6f75c3a89afe0cc81336cf0a2b89363a33c87e59b9fc/68747470733a2f2f66696c652e686f7573652f6f48435f2e706e67)
> screenshot_antag_icons_hereticsmuggler 	![](https://camo.githubusercontent.com/64f22a8166e3b1a3cf4f1c6fe9ff0bfd3d5dfe9ce51e8df348cfe28c7c3ff0d8/68747470733a2f2f66696c652e686f7573652f6a4d6f332e706e67) 	![](https://camo.githubusercontent.com/fa9fb8ee5c93e481b1c04c8db7d1c210c0a4342faa7a710939124d86c7f678eb/68747470733a2f2f66696c652e686f7573652f537673542e706e67) 	![](https://camo.githubusercontent.com/d5f74f459d77953e795f6f75c3a89afe0cc81336cf0a2b89363a33c87e59b9fc/68747470733a2f2f66696c652e686f7573652f6f48435f2e706e67)
> screenshot_antag_icons_loneoperative 	![](https://camo.githubusercontent.com/9c68d408bddbd904f99f239ebdb8fc7f9f63b4276a5140338b3631b1e1d5a1a8/68747470733a2f2f66696c652e686f7573652f656444422e706e67) 	![](https://camo.githubusercontent.com/af00e8c753f22ad6d5d8174bea4b96d98c4fd9f5317801e02a3de5472a1652ec/68747470733a2f2f66696c652e686f7573652f6b6650462e706e67) 	![](https://camo.githubusercontent.com/bf2cac5f14f35a90af53db33746bf39144ec6d3c6d8d2da8c89ba7b1a63d8f7a/68747470733a2f2f66696c652e686f7573652f52745a4e2e706e67)
> screenshot_antag_icons_nightmare 	![](https://camo.githubusercontent.com/81211bf6325a5582ebf840d5fd5972acc9f4bcc4fe347498d7e4c907d208994d/68747470733a2f2f66696c652e686f7573652f5f5f45462e706e67) 	![](https://camo.githubusercontent.com/c92a2a2704627ddd32495c83ee6b20f632c50b9801c13d8a299da4737bcdf5aa/68747470733a2f2f66696c652e686f7573652f35652d372e706e67) 	![](https://camo.githubusercontent.com/00d2f1cec82fe8f8a7d3b99a9b0656b5208884c2219a068a86b67e2e616dd9d5/68747470733a2f2f66696c652e686f7573652f4f6c74692e706e67)
> screenshot_antag_icons_obsessed 	![](https://camo.githubusercontent.com/b37d6c33dd819c13a5ef6506f1bc4c93886916ebb06da0dc0c96c63df0976351/68747470733a2f2f66696c652e686f7573652f515367542e706e67) 	![](https://camo.githubusercontent.com/00278e218b43264bf71f067b6e676fc19ed387e5148b4d53315e966e7788c163/68747470733a2f2f66696c652e686f7573652f446c77362e706e67) 	![](https://camo.githubusercontent.com/b900d57dd1f92c15d5b159fabd7174c60c6a3df2d2c203ab3abd6a6078be6405/68747470733a2f2f66696c652e686f7573652f632d45482e706e67)
> screenshot_antag_icons_operative 	![](https://camo.githubusercontent.com/eeb1b76cdabcad074997a05523bf4d0ee416dde9b075c6143dcc0bb1228e73c3/68747470733a2f2f66696c652e686f7573652f58694c352e706e67) 	![](https://camo.githubusercontent.com/6318f987844999db744b453cb5189169971889504d4be387013d5a0051439e67/68747470733a2f2f66696c652e686f7573652f523869562e706e67) 	![](https://camo.githubusercontent.com/dcef1890e3094d464958849cd0197b36466b28558edda24422a684fc53e30a1e/68747470733a2f2f66696c652e686f7573652f2d62436f2e706e67)
> screenshot_antag_icons_operativemidround 	![](https://camo.githubusercontent.com/eeb1b76cdabcad074997a05523bf4d0ee416dde9b075c6143dcc0bb1228e73c3/68747470733a2f2f66696c652e686f7573652f58694c352e706e67) 	![](https://camo.githubusercontent.com/6318f987844999db744b453cb5189169971889504d4be387013d5a0051439e67/68747470733a2f2f66696c652e686f7573652f523869562e706e67) 	![](https://camo.githubusercontent.com/dcef1890e3094d464958849cd0197b36466b28558edda24422a684fc53e30a1e/68747470733a2f2f66696c652e686f7573652f2d62436f2e706e67)
> screenshot_antag_icons_provocateur 	![](https://camo.githubusercontent.com/fd99a78aba94ede6faeb762db4b4b0542a28fa841123af6c0cb3ba38428edf70/68747470733a2f2f66696c652e686f7573652f69376b482e706e67) 	![](https://camo.githubusercontent.com/84690e6e6d568fd545e97a5831592e219d2ebb2bedfe82b125e1405136ef6d29/68747470733a2f2f66696c652e686f7573652f4a755a772e706e67) 	![](https://camo.githubusercontent.com/9a9a7392191d9043da05160cd800b84bf0432db95fd52afe6232900b2b7d6ec4/68747470733a2f2f66696c652e686f7573652f315568712e706e67)
> screenshot_antag_icons_spaceninja 	![](https://camo.githubusercontent.com/cfc608b108d6637981db9f6a4badbad66bbbbf41e4c0deaac6396a9ca58475a9/68747470733a2f2f66696c652e686f7573652f315661462e706e67) 	![](https://camo.githubusercontent.com/0adcb8d6f425c6e723da146b3da72651b4c9bf10b26b761cd6ccb5eaf4826a9b/68747470733a2f2f66696c652e686f7573652f366f56782e706e67) 	![](https://camo.githubusercontent.com/b35b36f604c675eaf1dde188713db48dbe83db9c723e533ce43bbb0fb9d30fd5/68747470733a2f2f66696c652e686f7573652f785f75742e706e67)
> screenshot_antag_icons_syndicateinfiltrator 	![](https://camo.githubusercontent.com/2e1f9134a109eba03ef3aa6619a80080c9a40eadb30b1082b9888e2aad6c6a8b/68747470733a2f2f66696c652e686f7573652f4a7045542e706e67) 	![](https://camo.githubusercontent.com/4984957759ccd866979ff0831f4ac5a12cd7f2174293fe8938787ad38e92ab09/68747470733a2f2f66696c652e686f7573652f543633652e706e67) 	![](https://camo.githubusercontent.com/dd79503dc0850177ce0656be20f5c6fadd7d6be1f4cdd299d78caa21a9950a54/68747470733a2f2f66696c652e686f7573652f7362386c2e706e67)
> screenshot_antag_icons_syndicatesleeperagent 	![](https://camo.githubusercontent.com/2e1f9134a109eba03ef3aa6619a80080c9a40eadb30b1082b9888e2aad6c6a8b/68747470733a2f2f66696c652e686f7573652f4a7045542e706e67) 	![](https://camo.githubusercontent.com/4984957759ccd866979ff0831f4ac5a12cd7f2174293fe8938787ad38e92ab09/68747470733a2f2f66696c652e686f7573652f543633652e706e67) 	![](https://camo.githubusercontent.com/dd79503dc0850177ce0656be20f5c6fadd7d6be1f4cdd299d78caa21a9950a54/68747470733a2f2f66696c652e686f7573652f7362386c2e706e67)
> screenshot_antag_icons_traitor 	![](https://camo.githubusercontent.com/2e1f9134a109eba03ef3aa6619a80080c9a40eadb30b1082b9888e2aad6c6a8b/68747470733a2f2f66696c652e686f7573652f4a7045542e706e67) 	![](https://camo.githubusercontent.com/4984957759ccd866979ff0831f4ac5a12cd7f2174293fe8938787ad38e92ab09/68747470733a2f2f66696c652e686f7573652f543633652e706e67) 	![](https://camo.githubusercontent.com/dd79503dc0850177ce0656be20f5c6fadd7d6be1f4cdd299d78caa21a9950a54/68747470733a2f2f66696c652e686f7573652f7362386c2e706e67)
> screenshot_antag_icons_wizard 	![](https://camo.githubusercontent.com/5e5e7cd89a77036fe0a8b4a915d0180357fbb6eafa408895ae5379a4588ed404/68747470733a2f2f66696c652e686f7573652f32595f5a2e706e67) 	![](https://camo.githubusercontent.com/ba028968fb9f579a4350ea18a128374144aabfcb385e326d46c5d0ac95c05921/68747470733a2f2f66696c652e686f7573652f586339442e706e67) 	![](https://camo.githubusercontent.com/757316a1cad8964f47afe471d45c356168a5997573028e0e56acdfc408783601/68747470733a2f2f66696c652e686f7573652f6e6b54312e706e67)
> screenshot_antag_icons_wizardmidround 	![](https://camo.githubusercontent.com/5e5e7cd89a77036fe0a8b4a915d0180357fbb6eafa408895ae5379a4588ed404/68747470733a2f2f66696c652e686f7573652f32595f5a2e706e67) 	![](https://camo.githubusercontent.com/ba028968fb9f579a4350ea18a128374144aabfcb385e326d46c5d0ac95c05921/68747470733a2f2f66696c652e686f7573652f586339442e706e67) 	![](https://camo.githubusercontent.com/757316a1cad8964f47afe471d45c356168a5997573028e0e56acdfc408783601/68747470733a2f2f66696c652e686f7573652f6e6b54312e706e67)
> ## Help
> What is this?
> I am changing sprites, it's supposed to look different.
> This is a false positive.

![image](https://user-images.githubusercontent.com/58055496/205223191-d71b3368-45e9-4441-8713-64ba436490d6.png)
works on my machine?

LemonInTheDark committed:
enables tracking

LemonInTheDark commented:
TMing this, will open a mergable pr in a sec

LemonInTheDark commented:
bro wtf

LemonInTheDark committed:
makes canisters hold off on visual updates unless something isactually changing

LemonInTheDark committed:
fuck

LemonInTheDark committed:
Fixed the funny haha black overlays

I forgot to have the "no turf exists" case default to the base plane. am
stupid, sorry :(

LemonInTheDark commented:
looking into it

LemonInTheDark commented:
miscalled update_appearance in the airlock mapping helper. I've fixed it.

LemonInTheDark committed:
Fixes up a miscalled update_appearance, fixes mapping helper'd airlocks not looking right on mapload
